test-runner-poc/lib/runner.js

```diff
@@ -29,18 +28,28 @@ function environmentSetup(environmentCallback, done) {
     createFilters(environmentParser);
     function environmentParser(environmentName, version) {
       const Environment = environments[environmentName];
-      client.db('admin').command({ isMaster: 1 }, (err, result) => {
+
+      function callEnvironmentCallback(result, version, err) {
+        const environment = new Environment(result, version);
+        environment.mongo = require('../../index');
+        environmentCallback(err, environment);
+        return client.close(done);
+      }
+
+      client.db('admin').command({ isMaster: 1, }, (err, result) => {
         if (err) environmentCallback(err)
-        for (let i = 0; i < result.hosts.length; i++) {
-          if (result.ismaster) {
-            const environment = new Environment(result, version);
-            environment.mongo = require('../../index');
-            environmentCallback(err, environment);
-            return client.close(done);
-          } else if (i === result.hosts.length - 1) {
-            return environmentParser(environmentName, version)
+        if (result.hosts) {
+          for (let i = 0; i < result.hosts.length; i++) {
+            if (result.ismaster) {
+              return callEnvironmentCallback(result, version, err);
```

Comment:
Will this loop complete if we are `return`ing for each value of `i`?

Reply:
The purpose of the `for` loop was to return once it locates the primary node, so it only needs to return once.

Reply:
If it reaches the else if statement then that means it has not found the primary node (the election hasn't taken place and all of the nodes are secondaries). In that case, it recursively calls `environmentParser` again.

Reply:
Okay, let's add some documentation to explain what's going on for future users.

Reply:
Added documentation